### DISCUSSION TITLE
PKCS11 codesigning in pipelines

### DISCUSSION THREAD

**Discussion by wak932** [airsdk/Adobe-Runtime-Support#2967] [Q&A]
Hoping someone has some experience of this. Since April, "industry changes" mean we aren't allowed to a .p12 file anymore from Digicert for codesigning using the adt command in our pipeline - we only found this out in the last few weeks and try as I might I cannot seem to get the set up correct for using their PKCS11 Keylocker service.

Very specifically we're using it in the adt command line, in a CircleCI windows instance. This is the current format of the command... 

adt.bat -package -storetype pkcs11 -storepass none -keystore "C:\Program Files\DigiCert\DigiCert Keylocker Tools\pkcs11properties.cfg"   -keypass <password> -providerName signingmanager -target bundle output deploy\\application.xml -C deploy .

but I always just get keystore type not available. I've installed the digicert tools and tested them from an ssh prompt - it all claims to be good. but no joy...
Have any other souls trodden this path and have wisdom to share?

**Comment by ajwfrost** [comment]
Hi

We've also had the issues of needing to change how we do code-signing on Windows - although for us, we just use a hardware dongle from Digicert which means we can still use the Windows "SignTool" to actually sign things, we just have to have the dongle plugged in and unlocked.

For AIR signatures, for what you're trying there, there are two parts to it:
1) initial AIR-specific code signing, which generates the "signatures.xml" file. This actually is primarily used by the AIR installer when you package up something as a .air file (and I think also as a .exe 'native' installer)
2) platform-specific code signing, which signs the executable files using the OS-specific mechanism

So your command-line above, if using pkcs11 store type and .p12 file, would have generated files in an "output" folder that were not actually signed according to Windows... To do that, you would have had to repeat the code-sign commands again after the `-target bundle` part and before the `output` name.

This means .. a potential workaround there is to just use a self-signed .p12 file for the AIR-specific signing, and then after you've created the bundle, use the Digicert tools in order to sign the generated .exe file (I'm hoping the "Adobe AIR.dll" file in the bundle's subfolder will be signed still with our own certificate).

I am currently in discussions with Digicert though and am hoping to work out a way in which we can use their set-up in a similar way to how the "KeychainStore" mechanism works on macOS.

Hope that helps..

thanks
   Andrew

**Comment by wak932** [comment]
thanks Andrew... that gives me something to go on. Good luck with Digicert - still waiting for contact back from them myself.

**Comment by ajwfrost** [comment]
Hi

Quick update .. I just found the below suggestion, from a while ago, for someone who may have been trying something similar:
https://github.com/airsdk/Adobe-Runtime-Support/issues/295#issuecomment-616742497
Would be interesting to know if this helped.

From some of the discussions I've had so far, it seems likely that some option might be possible, although it may require us to update how the Java code works within ADT... so e.g. looking at something like:
https://docs.digicert.com/en/software-trust-manager/signing-tools/sign-java-files-with-jarsigner-using-java-code-and-pkcs11-integration.html

The one thing I'm wondering: currently, in ADT, we ask for the store type and then for PKCS11, the "key store" and the certificate is just picked out of that (the default = alias "1", unless you specify that as an extra parameter). We've been adapting our own code-signing of Windows binaries (adl.exe etc) from our original mechanism - signtool.exe using a .p12 file and passcode - to the Digicert USB-dongle approach, and the adjustment we had to make was to pass in the certificate thumbprint. So potentially we need to have some option like that in order to get Java to provide us with the certificate and private key from this. 

thanks

**Reply by wak932** [reply]
Hi Andrew.

I looked at that MYWindows idea but came up against the fact we are a windows image on docker, so it did not list the key pair (Im guessing that was why).
I've also found plenty fo stuff on checking the provider is in the the java.security conf file. Certainly the SunPKCS provider is, and I've run a small java class, that seems to suggest the .cfg file I added is happily assigned to a provider.
`import java.security.Provider;
import java.security.Security;

public class LoadPKCS11Provider {
    public static void main(String[] args) throws Exception {
        String configName = "/Program Files/DigiCert/DigiCert Keylocker Tools/pkcs11properties.cfg";
        Provider p = Security.getProvider("SunPKCS11");
        p = p.configure(configName);
        if (p != null) {
            System.out.println("Provider loaded successfully"+ p.getName());
        } else {
            System.out.println("Failed to load provider");
        }
    }
}`

But still no joy. I've currently come back round to self-signing and then only using digicert with the msi.
